doc/glossary.rst

```diff
@@ -0,0 +1,968 @@
+.. currentmodule:: sklearn
+
+.. _glossary:
+
+=========================================
+Glossary of Common Terms and API Elements
+=========================================
+
+This glossary hopes to definitively represent the tacit and explicit
+conventions applied in Scikit-learn and its API, while providing a reference
+for users and contributors. It aims to describe the concepts and either detail
+their corresponding API or link to other relevant parts of the documentation
+which do so.
+
+General Concepts
+================
+
+.. glossary::
+
+    API
+        The specific interfaces for estimators implemented in Scikit-learn and
+        the generalized conventions across types of estimators as described in
+        this glossary.
+
+        The specific interfaces that constitute Scikit-learn's public API are
+        largely documented in :ref:`api_ref`. However we less formally consider
+        anything as public API if none of the identifiers required to access it
+        begins with ``_``.  We generally try to maintain :term:`backwards
+        compatibility` for all objects in the public API.
+
+        Private API, including functions, modules and methods beginning ``_``
+        are not assured to be stable.
+
+    array-like
+
+        The most common data format for *input* to Scikit-learn estimators and
+        functions, array-like is any type object for which
+        :func:`numpy.asarray` will produce an array of appropriate shape
+        (usually 1 or 2-dimensional) of appropriate dtype (usually numeric).
+
+        This includes:
+
+        * a numpy array
+        * a list of numbers
+        * a list of length-k lists of numbers for some fixed length k
+        * a :class:`pandas.DataFrame` with all columns numeric
```

Comment:
Seems that a link to pandas won't work in curret doc, maybe it worth to enable that in doc/conf.py?

Reply:
Done